src/thunderbolt/tbtctl.c

```diff
@@ -0,0 +1,615 @@
+                return EXIT_SUCCESS;
+        }
+
+        sl = tb_device_get_security_level(device);
+        if (sl < 0) {
+                log_error_errno(sl, "Failed to determine security level");
+                return EXIT_FAILURE;
+        }
+
+        if (sl != SECURITY_USER && sl != SECURITY_SECURE) {
+                log_debug("Security level of controller insufficient");
+                /* not an error here */
+                return EXIT_SUCCESS;
+        }
+
+        auth.level = MIN(auth.level, sl);
+
+        r = tb_device_authorize(device, &auth);
+        if (r < 0) {
+                log_error_errno(r, "Failed to authorize device: %m");
+                return EXIT_FAILURE;
+        }
+
+        return EXIT_SUCCESS;
+}
+
+static const struct CtlCmd cmd_udev = {
+        .name = "udev",
+        .func = authorize_udev,
+        .desc = "internal command for udev rules",
+        .root = true,
+};
+
+
+static int forget_device(struct udev *udev, int argc, char *argv[]) {
+        _cleanup_tb_store_free_ TbStore *store = NULL;
+        const char *uuid;
+        int r;
+
+        if (argc < 2) {
+                fprintf(stderr, "%s: need device uuid\n",
+                        program_invocation_short_name);
+                return EXIT_FAILURE;
+        }
+
+        uuid = argv[1];
+
+        r = tb_store_new(&store);
+        if (r < 0) {
+                log_error_errno(r, "Couldn't open store: %m");
+                return EXIT_FAILURE;
+        }
+
+        r = tb_store_remove_auth(store, uuid);
+        if (r < 0 && errno != -ENOENT) {
+                log_error_errno(r, "Could not remove authorization: %m");
+                return EXIT_FAILURE;
+        }
+
+        r = tb_store_remove_device(store, uuid);
+        if (r < 0) {
+                log_error_errno(r, "Could not remove device: %m");
+                return EXIT_FAILURE;
+        }
+
+        return EXIT_SUCCESS;
+}
+
+static const struct CtlCmd cmd_forget = {
+        .name = "forget",
+        .func = forget_device,
+        .desc = "Remove a device from the database",
+        .root = true,
+};
+
+
+static const struct CtlCmd *ctrl_cmds[] = {
+        &cmd_list,
+        &cmd_authorize,
+        &cmd_forget,
+
+        &cmd_udev
+};
+
+static void help(void) {
+        unsigned int i;
+
+        printf("%s [--version] [--debug] COMMAND [OPTIONS]\n\n"
+               "Manager thunderbolt devices\n\n"
+               "  -h --help             Show this help and exit\n"
+               "  --version             Print version string and exit\n"
+               "\n"
+               "Commands:\n"
+               , program_invocation_short_name);
+
+        for (i = 0; i < ELEMENTSOF(ctrl_cmds); i++) {
+                const struct CtlCmd *cmd = ctrl_cmds[i];
+                if (!cmd->desc)
+                        continue;
+
+                printf("  %-20s  %s\n", cmd->name, cmd->desc);
+        }
+}
+
+
+#define ARG_VERSION  0x100
+#define ARG_NOROOT   0x101
```

Comment:
why use macros for this, when anonymous enums work too? we generally prefer proper C constructs over pre-processor constructs, if both are easy and equivalent